chain/stmgr/forks.go

```diff
@@ -41,8 +41,11 @@ type MigrationCache interface {
 // - The oldState is the state produced by the upgrade epoch.
 // - The returned newState is the new state that will be used by the next epoch.
 // - The height is the upgrade epoch height (already executed).
-// - The tipset is the tipset for the last non-null block before the upgrade. Do
-//   not assume that ts.Height() is the upgrade height.
+// - The tipset is the first non-null tipset after the upgrade height (the tipset in
+//   which the upgrade is executed). Do not assume that ts.Height() is the upgrade height.
+//
+// NOTE: In StateCompute and CallWithGas, the passed tipset is actually the tipset _before_ the
```

Comment:
Do the historical upgrades all work if we consistently pass in the tipset before the upgrade? It looks like a lot of the recent ones don't use this param at all. We should file an issue for checking this and if so refactoring the upgrade functions to take in this tipset instead so that Call and ComputeState never have access to weird state.

Reply:
As far as I can tell, it's used for lookbacks. But it may also be used for randomness in some cases?

Reply:
https://github.com/filecoin-project/lotus/issues/7226